documentation/src/main/asciidoc/mapper-orm-indexing-massindexer.asciidoc

```diff
@@ -0,0 +1,257 @@
+but there is no one-size-fits-all solution,
+so some configuration is required to get the best of it.
+
+Performance optimization can get quite complex,
+so keep the following in mind while you attempt to configure the `MassIndexer`:
+
+* Always test your changes to assess their actual effect:
+advice provided in this section is true in general,
+but each application and environment is different,
+and some options, when combined, may produce unexpected results.
+* Take baby steps:
+before tuning mass indexing with 40 indexed entity types
+with two million instances each,
+try a more reasonable scenario with only one entity type,
+optionally limiting the number of entities to index
+to assess performance more quickly.
+* Tune your entity types individually *before*
+you try to tune a mass indexing operation that indexes multiple entity types in parallel.
+
+[[mapper-orm-indexing-massindexer-tuning-threads]]
+=== Threads and JDBC connections
+// Search 5 anchors backward compatibility
+[[search-batchindexing-threadsandconnections]]
+
+Increasing parallelism usually helps as the bottleneck
+usually is the latency to the database connection:
+it's probably worth it to experiment with a number of threads significantly higher
+than the number of actual cores available.
+
+However, each thread requires one JDBC connection,
+and JDBC connections are usually in limited supply.
+In order to increase the number of threads safely:
+
+1. You should make sure your database can actually handle the resulting number of connections.
+2. Your JDBC connection pool should be configured to provide a sufficient number of connections.
+3. The above should take into account the rest of your application (request threads in a web application):
+ignoring this may bring other processes to a halt while the `MassIndexer` is working.
+
+There is a simple formula to understand how
+the different options applied to the `MassIndexer` affect
+the number of used worker threads and connections:
+
+----
+threads = typesToIndexInParallel * (threadsToLoadObjects + 1);
+required JDBC connections = threads;
+----
+
+Here are a few suggestions for a roughly sane tuning starting point
+for the parameters that affect parallelism:
+
+`typesToIndexInParallel`::
+Should probably be a low value, like 1 or 2, depending on how much
+of your CPUs have spare cycles and how slow a database round trip will be.
+`threadsToLoadObjects`::
+Higher increases the pre-loading rate for the picked entities from
+the database, but also increases memory usage and the pressure on the threads working on subsequent
+indexing.
+Note that each thread will extract data from the entity to reindex,
+which depending on your mapping might require to access lazy associations
+and load associated entities, thus making blocking calls to the database,
+so you will probably need a high number of threads working in parallel.
+
+[TIP]
+====
+All internal thread groups have meaningful names,
```

Comment:
That's nice for tracing. Maybe we can quote their names / or the name pattern here.

Reply:
I added a mention of the common prefix ("Hibernate Search") but I'd rather not go further, because the thread names are likely to change in future releases and I'm sure we'll forget about this part of the documentation.